[Bug][pydevlake] Panic when running blueprint with large-ish amounts of data

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar issues.


### What happened

We have an internal release tool which I've written a plugin to scrape releases from. We're using the new `pydevlake` plugin stuff.

I've obfuscated some of the logs because the details aren't relevant to outside of where I work so I hope I've not removed anything useful.

When trigger a run of a blueprint (`curl -v -X 'POST' 'http://localhost:8080/blueprints/1/trigger'`) we end up with a `panic`:

```
...
time="2023-03-24 15:01:22" level=info msg=" [pipeline service] [pipeline #1] [task #1] [collectHopperReleases] DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=13 HTTP/1.1" 200 None"
time="2023-03-24 15:01:22" level=info msg=" [pipeline service] [pipeline #1] [task #1] [collectHopperReleases] DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=15 HTTP/1.1" 200 None"
time="2023-03-24 15:01:22" level=info msg=" [pipeline service] [pipeline #1] [task #1] [collectHopperReleases] DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=17 HTTP/1.1" 200 None"
time="2023-03-24 15:01:23" level=info msg=" [pipeline service] [pipeline #1] [task #1] [collectHopperReleases] DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=19 HTTP/1.1" 200 None"
time="2023-03-24 15:01:23" level=info msg=" [pipeline service] [pipeline #1] [task #1] [collectHopperReleases] DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=21 HTTP/1.1" 200 None"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x10082d0aa]

goroutine 77 [running]:
github.com/apache/incubator-devlake/core/utils.scanOutputPipe.func1()
        /Users/sam/src/incubator-devlake/backend/core/utils/ipc.go:246 +0xea
created by github.com/apache/incubator-devlake/core/utils.StreamProcess
        /Users/sam/src/incubator-devlake/backend/core/utils/ipc.go:190 +0x34b
exit status 2
make[1]: *** [run] Error 1
make: *** [dev] Error 2
```

This works fine when run directly:

```
poetry run hopper/main.py collect "{\"db_url\": \"mysql://merico:merico@127.0.0.1:3306/lake\", \"scope\": {\"id\": \"test-http\", \"name\": \"test-http\"}, \"connection\": { \"id\": 1, \"name\": \"hopper\", \"endpoint\": \"https://***.net\", \"token\": \"${TOKEN}\"}}" releases 3>&1
DEBUG: Starting new HTTPS connection (1): ***.net:443
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=1 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=3 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=5 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=7 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=9 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=11 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=13 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=15 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=17 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=19 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=21 HTTP/1.1" 200 None
{"increment": 100, "current": 100}
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=23 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=23 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=23 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=25 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=27 HTTP/1.1" 200 None
DEBUG: https://***.net:443 "GET /api/apps/test-http/releases?page=29 HTTP/1.1" 200 None
...
```

I suspect what happens is [the message sent here](https://github.com/apache/incubator-devlake/blob/5f89c21181ca49cd10253078665a9124e030469a/backend/python/pydevlake/pydevlake/subtasks.py#L68-L71) isn't correctly handled on the go side of the IPC.

I've worked around this in the short term by setting `sync_point_interval` to something very large.

### What do you expect to happen

It should not panic the blueprint should run successfully.

### How to reproduce

See above.

### Anything else

Every time.

### Version

main

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/backend/core/utils/ipc.go
+++ b/backend/core/utils/ipc.go
@@ -243,7 +243,9 @@ func scanOutputPipe(pipe io.ReadCloser, wg *sync.WaitGroup, onReceive func([]byt
 			src := scanner.Bytes()
 			data := make([]byte, len(src))
 			copy(data, src)
-			onReceive(data)
+			if onReceive != nil {
+				onReceive(data)
+			}
 			outboundChannel <- responseCreator(data)
 		}
 		wg.Done()
@@ -259,7 +261,9 @@ func scanErrorPipe(pipe io.ReadCloser, onReceive func([]byte), outboundChannel c
 			src := scanner.Bytes()
 			data := make([]byte, len(src))
 			copy(data, src)
-			onReceive(data)
+			if onReceive != nil {
+				onReceive(data)
+			}
 			outboundChannel <- &ProcessResponse{stderr: data}
 			_, _ = remoteErrorMsg.Write(src)
 			_, _ = remoteErrorMsg.WriteString("\n")

--- a/backend/go.mod
+++ b/backend/go.mod
@@ -68,7 +68,7 @@ require (
 	github.com/go-openapi/swag v0.21.1 // indirect
 	github.com/go-playground/locales v0.14.0 // indirect
 	github.com/go-playground/universal-translator v0.18.0 // indirect
-	github.com/go-sql-driver/mysql v1.6.0
+	github.com/go-sql-driver/mysql v1.6.0 // indirect
 	github.com/gogo/googleapis v1.4.1 // indirect
 	github.com/gogo/protobuf v1.3.2 // indirect
 	github.com/gogo/status v1.1.0 // indirect

--- a/backend/python/pydevlake/pydevlake/__init__.py
+++ b/backend/python/pydevlake/pydevlake/__init__.py
@@ -22,3 +22,9 @@ from .message import RemoteScopeGroup
 from .plugin import Plugin, ScopeTxRulePair
 from .stream import DomainType, Stream, Substream
 from .context import Context
+
+# the debugger hangs on startup during plugin registration (reason unknown), hence this workaround
+import sys
+if not sys.argv.__contains__('startup'):
+    from pydevlake.helpers import debugger
+    debugger.init()

--- a/backend/python/pydevlake/pydevlake/helpers/debugger.py
+++ b/backend/python/pydevlake/pydevlake/helpers/debugger.py
@@ -18,7 +18,7 @@ import os
 from pydevlake import logger
 
 
-def __start__():
+def init():
     debugger = os.getenv("USE_PYTHON_DEBUGGER", default="").lower()
     if debugger == "":
         return
@@ -32,12 +32,10 @@ def __start__():
             import pydevd_pycharm as pydevd
             try:
                 pydevd.settrace(host=host, port=port, suspend=False, stdoutToServer=True, stderrToServer=True)
+                logger.info("Pycharm remote debugger successfully connected")
             except TimeoutError as e:
                 logger.error(f"Failed to connect to pycharm debugger on {host}:{port}. Make sure it is running")
         except ImportError as e:
             logger.error("Pycharm debugger library is not installed")
     else:
         logger.error(f"Unsupported Python debugger specified: {debugger}")
-
-
-__start__()

--- a/backend/python/pydevlake/pydevlake/ipc.py
+++ b/backend/python/pydevlake/pydevlake/ipc.py
@@ -24,9 +24,6 @@ from pydevlake.message import Message
 
 
 def plugin_method(func):
-    # keep this to enable debugging, and don't move this elsewhere or it can cause crashes if it gets called during the bootstrap process.
-    # noinspection PyUnresolvedReferences
-    from pydevlake.helpers import debugger
 
     def open_send_channel() -> TextIO:
         fd = 3

--- a/backend/python/pydevlake/pydevlake/subtasks.py
+++ b/backend/python/pydevlake/pydevlake/subtasks.py
@@ -4,7 +4,7 @@
 # The ASF licenses this file to You under the Apache License, Version 2.0
 # (the "License"); you may not use this file except in compliance with
 # the License.  You may obtain a copy of the License at
-import time
+
 #     http://www.apache.org/licenses/LICENSE-2.0
 
 # Unless required by applicable law or agreed to in writing, software
@@ -45,7 +45,7 @@ class Subtask:
     def verb(self) -> str:
         pass
 
-    def run(self, ctx: Context, sync_point_interval=1):
+    def run(self, ctx: Context, sync_point_interval=100):
         with Session(ctx.engine) as session:
             subtask_run = self._start_subtask(session, ctx.connection.id)
             if ctx.incremental:
@@ -57,14 +57,11 @@ class Subtask:
             try:
                 for i, (data, state) in enumerate(self.fetch(state, session, ctx)):
                     self.process(data, session, ctx)
-                    # print("yielding python progress")
                     if i % sync_point_interval == 0 and i != 0:
-
                         # Save current state
                         subtask_run.state = json.dumps(state)
                         session.merge(subtask_run)
                         session.commit()
-                        time.sleep(1)
                         # Send progress
                         yield RemoteProgress(
                             increment=sync_point_interval,

--- a/backend/python/pydevlake/pydevlake/subtasks.py
+++ b/backend/python/pydevlake/pydevlake/subtasks.py
@@ -55,9 +55,12 @@ class Subtask:
                 state = dict()
 
             try:
-                for i, (data, state) in enumerate(self.fetch(state, session, ctx)):
+                records = self.fetch(state, session, ctx)
+                progress = last_progress = 0
+                for data, state in records:
+                    progress += 1
                     self.process(data, session, ctx)
-                    if i % sync_point_interval == 0 and i != 0:
+                    if progress % sync_point_interval == 0:
                         # Save current state
                         subtask_run.state = json.dumps(state)
                         session.merge(subtask_run)
@@ -65,8 +68,15 @@ class Subtask:
                         # Send progress
                         yield RemoteProgress(
                             increment=sync_point_interval,
-                            current=i
+                            current=progress
                         )
+                        last_progress = progress
+                # Send final progress
+                if progress != last_progress:
+                    yield RemoteProgress(
+                        increment=progress-last_progress,
+                        current=progress
+                    )
             except Exception as e:
                 logger.error(f'{type(e).__name__}: {e}')
                 raise e
@@ -97,7 +107,7 @@ class Subtask:
         pass
 
     @abstractmethod
-    def process(self, data: object, session: Session):
+    def process(self, data: object, session: Session, ctx: Context):
         """
         Called for all data entries returned by `fetch`.
         """

--- a/backend/python/pydevlake/pydevlake/subtasks.py
+++ b/backend/python/pydevlake/pydevlake/subtasks.py
@@ -4,7 +4,7 @@
 # The ASF licenses this file to You under the Apache License, Version 2.0
 # (the "License"); you may not use this file except in compliance with
 # the License.  You may obtain a copy of the License at
-
+import time
 #     http://www.apache.org/licenses/LICENSE-2.0
 
 # Unless required by applicable law or agreed to in writing, software
@@ -45,7 +45,7 @@ class Subtask:
     def verb(self) -> str:
         pass
 
-    def run(self, ctx: Context, sync_point_interval=100):
+    def run(self, ctx: Context, sync_point_interval=1):
         with Session(ctx.engine) as session:
             subtask_run = self._start_subtask(session, ctx.connection.id)
             if ctx.incremental:
@@ -57,13 +57,14 @@ class Subtask:
             try:
                 for i, (data, state) in enumerate(self.fetch(state, session, ctx)):
                     self.process(data, session, ctx)
-
+                    # print("yielding python progress")
                     if i % sync_point_interval == 0 and i != 0:
+
                         # Save current state
                         subtask_run.state = json.dumps(state)
                         session.merge(subtask_run)
                         session.commit()
-
+                        time.sleep(1)
                         # Send progress
                         yield RemoteProgress(
                             increment=sync_point_interval,

--- a/backend/python/test/fakeplugin/fakeplugin/main.py
+++ b/backend/python/test/fakeplugin/fakeplugin/main.py
@@ -23,7 +23,6 @@ from sqlmodel import Field
 from pydevlake import Plugin, Connection, TransformationRule, Stream, ToolModel, ToolScope, RemoteScopeGroup, DomainType
 from pydevlake.domain_layer.devops import CicdScope, CICDPipeline, CICDStatus, CICDResult, CICDType
 
-
 VALID_TOKEN = "this_is_a_valid_token"
 
 
@@ -40,18 +39,12 @@ class FakePipeline(ToolModel, table=True):
     state: State
 
 
-class FakeStream(Stream):
+class FakePipelineStream(Stream):
     tool_model = FakePipeline
     domain_types = [DomainType.CICD]
 
-    fake_pipelines = [
-        FakePipeline(id=1, state=FakePipeline.State.SUCCESS, started_at=datetime(2023, 1, 10, 11, 0, 0), finished_at=datetime(2023, 1, 10, 11, 3, 0)),
-        FakePipeline(id=2, state=FakePipeline.State.FAILURE, started_at=datetime(2023, 1, 10, 12, 0, 0), finished_at=datetime(2023, 1, 10, 12, 1, 30)),
-        FakePipeline(id=1, state=FakePipeline.State.PENDING),
-    ]
-
     def collect(self, state, context):
-        for p in self.fake_pipelines:
+        for p in self.generate_fake_pipelines():
             yield json.loads(p.json()), {}
 
     def convert(self, pipeline: FakePipeline, ctx):
@@ -89,6 +82,19 @@ class FakeStream(Stream):
         if pipeline.finished_at:
             return (pipeline.finished_at - pipeline.started_at).seconds
         return None
+
+    @classmethod
+    def generate_fake_pipelines(cls) -> list[FakePipeline]:
+        states = [FakePipeline.State.SUCCESS, FakePipeline.State.FAILURE, FakePipeline.State.PENDING]
+        fake_pipelines = []
+        for i in range(250):
+            fake_pipelines.append(FakePipeline(
+                id=i,
+                state=states[i % len(states)],
+                started_at=datetime(2023, 1, 10, 11, 0, 0, microsecond=i),
+                finished_at=datetime(2023, 1, 10, 11, 3, 0, microsecond=i),
+            ))
+        return fake_pipelines
 
 
 class FakeConnection(Connection):
@@ -149,7 +155,7 @@ class FakePlugin(Plugin):
     @property
     def streams(self):
         return [
-            FakeStream
+            FakePipelineStream
         ]
 
 

--- a/backend/server/services/remote/bridge/bridge.go
+++ b/backend/server/services/remote/bridge/bridge.go
@@ -53,7 +53,7 @@ func (b *Bridge) RemoteSubtaskEntrypointHandler(subtaskMeta models.SubtaskMeta) 
 			if err != nil {
 				return err
 			}
-			if progress.Current != 0 {
+			if progress.Total != 0 {
 				ctx.SetProgress(progress.Current, progress.Total)
 			} else if progress.Increment != 0 {
 				ctx.IncProgress(progress.Increment)

